### PR TITLE
Expose the deployment of SQL CLI workflows in the Astro CLI

### PR DESCRIPTION
**Pull request by pankajkoti**
## Description

>Implement the command astro flow deploy in the Astro CLI. As a starting point, this command will work on Astro Cloud and it will deploy using the image-based approach.
Also, Identify the Python SDK availability/version compatibility and promote the correct version of Python SDK.
Creates environment specific dags folders `<project-dir>/.deploy/<env>/dags` and deploys DAGs from those directory to Astro Cloud by successively calling the `astro deploy` command post taking care of SQL CLI related deploy implementation.

**WIP**: Unit tests and tidying up the code.

## 🎟 Issue(s)

closes: https://github.com/astronomer/astro-sdk/issues/1473
closes: https://github.com/astronomer/astro-sdk/issues/1520
closes: https://github.com/astronomer/astro-sdk/issues/1602

## 🧪 Functional Testing

> List the functional testing steps to confirm this feature or fix.

## 📸 Screenshots

> Add screenshots to illustrate the validity of these changes.

## 📋 Checklist

- [ ] Rebased from the main (or release if patching) branch (before testing)
- [ ] Ran `make test` before taking out of draft
- [ ] Ran `make lint` before taking out of draft
- [ ] Added/updated applicable tests
- [ ] Tested against [Astro-API](https://github.com/astronomer/astro/) (if necessary).
- [ ] Tested against [Houston-API](https://github.com/astronomer/houston-api/) and [Astronomer](https://github.com/astronomer/astronomer/) (if necessary).
- [ ] Communicated to/tagged owners of respective clients potentially impacted by these changes.
- [ ] Updated any related [documentation](https://github.com/astronomer/docs/)

### CODE REVIEW

File: cmd/sql/flow.go
Comment (tatiana) [inline cmd/sql/flow.go:358]: Is there any scenario where retrieving the latest version in PyPI would not match the local version of the SQL CLI?

File: sql/flow_config.go
Comment (tatiana) [inline sql/flow_config.go:23]: @pankajkoti what do you think if we check both for the lower and upper bound versions?
https://github.com/astronomer/astro-sdk/issues/1473

```
            "astroSDKPythonMaxVersion": "",
            "astroSDKPythonMinVersion": "1.3.1",
            "astroRuntimeMinVersion": "6.1.0",
            "astroRuntimeMaxVersion": "7.1.0"
```

This would be helpful if we need to introduce breaking changes in the future.

File: cmd/sql/flow.go
Comment (tatiana) [inline cmd/sql/flow.go:362]: To be future-proof, we can have a lower and upper threshold of supported versions, as mentioned in the comment
https://github.com/astronomer/astro-cli/pull/1007/files#r1063343197

File: cmd/sql/flow.go
Comment (tatiana) [inline cmd/sql/flow.go:374]: We should be explicit that if users do not add this dependency, the deployment will not proceed.

File: cmd/sql/flow.go
Comment (tatiana) [inline cmd/sql/flow.go:389]: What happens if the user has an older version of the Python SDK? Should we prompt again and delete the old version?

File: cmd/sql/flow.go
Comment (pankajkoti) [reply cmd/sql/flow.go:358]: The local version of SQL CLI is always the latest at the moment.

File: sql/flow_config.go
Comment (feluelle) [reply sql/flow_config.go:23]: Maybe we can have sth. like `>6.1.0,<7.1.0` syntax as in python. Maybe go-version supports that?

File: cloud/deploy/deploy.go
Comment (tatiana) [inline cloud/deploy/deploy.go:106]: WDYT of naming this as `dagsFolderPath`?

File: cmd/cloud/deploy.go
Comment (tatiana) [inline cmd/cloud/deploy.go:68]: Should we hide this flag?
What happens if users run `astro deploy --help`? Are we displaying it?

File: cmd/sql/flow.go
Comment (neel-astro) [inline cmd/sql/flow.go:625]: please check once the output for `astro flow deploy --help`, I don't think so flow library which handles help output for other commands will work for this one, in which case we would need the default help out with description of these flags being set

File: go.sum
Comment (neel-astro) [inline go.sum:653]: please run `go mod tidy` to clean up the `go.sum` file of unused dependencies

File: sql/flow.go
Comment (neel-astro) [inline sql/flow.go:21]: I am guessing this is temporary, leaving a comment for our reference to remove it before merging the PR

File: cmd/sql/flow.go
Comment (neel-astro) [inline cmd/sql/flow.go:373]: nit: ideally I would expect most of the logic inside this function to be part of `sql` package instead of `cmd/sql`. In my mind this is how I see the SQL code structure at a high level:
`cmd/sql` -> anything related to command declaration for all the SQL CLI commands
`sql` -> all the business logic outside the `cobra` command declaration for all the SQL CLI commands

File: sql/flow_config.go
Comment (neel-astro) [inline sql/flow_config.go:93]: nit: this should go into the global var block, and be reused in all the methods of the package. There is no point in creating a new HTTP client for each individual method.

File: cmd/sql/flow.go
Comment (neel-astro) [inline cmd/sql/flow.go:456]: nit: should we be calling `cloud.Deploy` instead of executing the deploy command, I feel that to be more correct implementation in my mind 🤔, will leave it up to you to take a final call

**Review by neel-astro — COMMENTED** [review]
left bunch of nitpicks, looks fine otherwise

File: sql/flow_config.go
Comment (pankajkoti) [reply sql/flow_config.go:23]: Handling as per Felix's suggestions.

File: cloud/deploy/deploy.go
Comment (pankajkoti) [reply cloud/deploy/deploy.go:106]: keeping it to dagsPath as it being referred to so in the existing code for `astro deploy` and would not like to rename it in this PR.

File: cmd/sql/flow.go
Comment (pankajkoti) [reply cmd/sql/flow.go:410]: Since we append the required version to the end of file, currently it may not be an issue. Let's discuss this over a call together with Felix. Felix and I had a preliminary discussion on this and need to conclude together with you.

File: cmd/sql/flow.go
Comment (pankajkoti) [reply cmd/sql/flow.go:456]: @feluelle and I discussed this and feel that keeping the existing approach is suitable since by initialising the cloud.NewDeployCmd() we have it's default flag values and need not worry about constructing the deploy command otherwise that would need to be pass to cloud.Deploy. And the ExecuteC() method would know what to call it with once we have the cloud.NewDeployCmd() initialised.

File: cmd/sql/flow.go
Comment (tatiana) [reply cmd/sql/flow.go]: @pankajkoti I had a discussion with @feluelle last week, which will make this no longer true - shall we update the ticket description for https://github.com/astronomer/astro-sdk/issues/1582, so we improve this in that ticket?

File: cmd/sql/flow.go
Comment (pankajkoti) [reply cmd/sql/flow.go]: Log a ticket to enhance this post 0.3

File: cmd/sql/flow.go
Comment (feluelle) [reply cmd/sql/flow.go:556]: Thank you for the heads-up. This https://github.com/astronomer/astro-sdk/pull/1615 should handle that :)

File: pkg/input/input.go
Comment (neel-astro) [inline pkg/input/input.go:61]: we should be adding unit tests for these functions since it has been added to `pkg` which is a shared package.

**Review by neel-astro — COMMENTED** [review]
overall looks fine to me, have left few minor comments

File: cmd/sql/flow.go
Comment (pankajkoti) [reply cmd/sql/flow.go:85]: @feluelle ☝🏽 :)

File: pkg/input/input.go
Comment (pankajkoti) [reply pkg/input/input.go:61]: @neel-astro would appreciate you help here, quite unsure how to add unit tests for these components.

File: sql/flow.go
Comment (pankajkoti) [reply sql/flow.go:245]: This will be controlled by the compatibility matrix. It may not be always pinned to a specific version. It could also be of the format `>=2.1.0, <3.0.0` which is configured here: https://github.com/astronomer/astro-sdk/blob/astro-cli/sql-cli/config/astro-cli.json

File: cmd/sql/flow.go
Comment (feluelle) [reply cmd/sql/flow.go:85]: In my opinion cognitive complexity can't always be measured based on number of lines hence the `nolint`. It is quite straightforward code. Refactoring this in my opinion only adds more boilerplate code.

Comment (feluelle) [reply cmd/sql/flow.go:85]: @neel-astro I'm open to suggestions for refactoring this.

File: sql/flow.go
Comment (pankajkoti) [reply sql/flow.go:217]: checking that the array has minimum 2 items.

File: pkg/input/input.go
Comment (kushalmalani) [inline pkg/input/input.go:66]: Does this have to be case sensitive? Also, should there be a default response if user doesnt enter anything. 

Since this a common package, we should allow for those options to be configured

File: pkg/input/input.go
Comment (kushalmalani) [inline pkg/input/input.go:44]: nit: typo change displayig to displaying

**Review by kushalmalani — APPROVED** [review]
Dag deploy part of the code looks good to me. Left a comment for the common package (input) function. Will Lean on SQL team folks to review SQL code changes.

File: pkg/input/input.go
Comment (neel-astro) [reply pkg/input/input.go:61]:
```Golang
func TestPromptGetConfirmation(t *testing.T) {
	runner := GetYesNoSelector(PromptContent{Label: "test label, enter y/n"})
	runner.Keys = &promptui.SelectKeys{Next: promptui.Key{Code: rune('S')}, Prev: promptui.Key{Code: rune('W')}, PageUp: promptui.Key{Code: rune('D')}, PageDown: promptui.Key{Code: rune('A')}}
	tests := []struct {
		name        string
		inputString string
		want        bool
		wantErr     bool
	}{
		{
			name:        "basic yes case",
			inputString: "\n",
			want:        true,
			wantErr:     false,
		},
		{
			name:        "basic no case",
			inputString: "S\n",
			want:        false,
			wantErr:     false,
		},
		{
			name:        "no input case",
			inputString: "",
			want:        false,
			wantErr:     true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			runner.Stdin = io.NopCloser(strings.NewReader(tt.inputString))
			got, err := PromptGetConfirmation(runner)
			if (err != nil) != tt.wantErr {
				t.Errorf("PromptGetConfirmation() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if got != tt.want {
				t.Errorf("PromptGetConfirmation() = %v, want %v", got, tt.want)
			}
		})
	}
}
```

File: cmd/sql/flow.go
Comment (feluelle) [reply cmd/sql/flow.go]: Thank you. I have pushed the changes. PTAL again.

File: pkg/input/input.go
Comment (feluelle) [reply pkg/input/input.go:61]: Added in b3bd5e2

File: pkg/input/input.go
Comment (feluelle) [reply pkg/input/input.go:44]:
```suggestion
// Structure to hold content required for displaying prompts required for promptui library functions
```

File: pkg/input/input.go
Comment (feluelle) [reply pkg/input/input.go:66]: That's right, but we tried to be consistent with `Confirm`:
```golang
// Confirm requests a user to confirm their input
func Confirm(promptText string) (bool, error) {
	reader := bufio.NewReader(os.Stdin)
	fmt.Printf("%s (y/n) ", promptText)

	text, _ := reader.ReadString('\n')
	return strings.Trim(text, "\r\n") == "y", nil
}
```

I would say we can improve that separately sometime - and consistently for all inputs.

Comment (feluelle) [reply pkg/input/input.go:66]: We are also planning to add a non-prompt way separately for the prompt in our flow deploy command.

**Review by neel-astro — APPROVED** [review]

**Pull request merged**
